### PR TITLE
test: add dangling error e2e test

### LINKED ISSUE #500: add dangling error e2e test
**Is your feature request related to a problem?/Why is this needed**

**Describe the solution you'd like in detail**

https://github.com/kubernetes-sigs/azuredisk-csi-driver/pull/499 added the dangling error support for this CSI driver, need to also add a dangling error e2e test, below are the steps to validate dangling error:

Below are the steps how to validate dangling error support:
 - create a disk pvc
 - attach disk to node#1 manually, and `kubectl cordon node#1`
 - create a new pod with disk pvc mount, and pow would be scheduled to other node(e.g. node#2) and started successfullly
 - check pod events and azuredisk-controller logs

```
Events:
  Type    Reason                  Age   From                                        Message
  ----    ------                  ----  ----                                        -------
  Normal  Scheduled               52s   default-scheduler                           Successfully assigned default/statefulset-azuredisk-0 to aks-agentpool-60632172-vmss000007
  Normal  SuccessfulAttachVolume  27s   attachdetach-controller                     AttachVolume.Attach succeeded for volume "pvc-64ff3106-d091-45b0-ab0b-0f99c4bcb491"
  Normal  Pulling                 16s   kubelet, aks-agentpool-60632172-vmss000007  Pulling image "nginx"
  Normal  Pulled                  13s   kubelet, aks-agentpool-60632172-vmss000007  Successfully pulled image "nginx"
  Normal  Created                 10s   kubelet, aks-agentpool-60632172-vmss000007  Created container statefulset-azuredisk
  Normal  Started                 10s   kubelet, aks-agentpool-60632172-vmss000007  Started container statefulset-azuredisk


I0823 14:54:03.202098       1 utils.go:108] GRPC call: /csi.v1.Controller/ControllerPublishVolume
I0823 14:54:03.202118       1 utils.go:109] GRPC request: {"node_id":"aks-agentpool-60632172-vmss000007","volume_capability":{"AccessType":{"Mount":{"fs_type":"ext4"}},"access_mode":{"mode":1}},"volume_context":{"skuname":"StandardSSD_LRS","storage.kubernetes.io/csiProvisionerIdentity":"1598193782156-1847-disk.csi.azure.com"},"volume_id":"/subscriptions/xxx/resourceGroups/mc_andy-aks11510_andy-aks11510_eastus2/providers/Microsoft.Compute/disks/pvc-64ff3106-d091-45b0-ab0b-0f99c4bcb491"}
I0823 14:54:03.244538       1 controllerserver.go:392] GetDiskLun returned: <nil>. Initiating attaching volume "/subscriptions/xxx/resourceGroups/mc_andy-aks11510_andy-aks11510_eastus2/providers/Microsoft.Compute/disks/pvc-64ff3106-d091-45b0-ab0b-0f99c4bcb491" to node "aks-agentpool-60632172-vmss000007".
I0823 14:54:03.244609       1 controllerserver.go:409] Trying to attach volume "/subscriptions/xxx/resourceGroups/mc_andy-aks11510_andy-aks11510_eastus2/providers/Microsoft.Compute/disks/pvc-64ff3106-d091-45b0-ab0b-0f99c4bcb491" to node "aks-agentpool-60632172-vmss000007"
I0823 14:54:03.271050       1 azure_vmss.go:253] Couldn't find VMSS VM with scaleSetName "aks-agentpool-60632172-vmss" and instanceID "6", refreshing the cache
I0823 14:54:03.362862       1 azure_controller_common.go:152] found dangling volume /subscriptions/xxx/resourceGroups/mc_andy-aks11510_andy-aks11510_eastus2/providers/Microsoft.Compute/disks/pvc-64ff3106-d091-45b0-ab0b-0f99c4bcb491 attached to node aks-agentpool-60632172-vmss000006
W0823 14:54:03.362898       1 controllerserver.go:416] volume "/subscriptions/xxx/resourceGroups/mc_andy-aks11510_andy-aks11510_eastus2/providers/Microsoft.Compute/disks/pvc-64ff3106-d091-45b0-ab0b-0f99c4bcb491" is already attached to node "aks-agentpool-60632172-vmss000006", try detach first
I0823 14:54:03.362950       1 azure_controller_common.go:208] detach /subscriptions/xxx/resourceGroups/mc_andy-aks11510_andy-aks11510_eastus2/providers/Microsoft.Compute/disks/pvc-64ff3106-d091-45b0-ab0b-0f99c4bcb491 from node "aks-agentpool-60632172-vmss000006"
I0823 14:54:03.362966       1 azure_controller_vmss.go:149] azureDisk - detach disk: name "pvc-64ff3106-d091-45b0-ab0b-0f99c4bcb491" uri "/subscriptions/xxx/resourceGroups/mc_andy-aks11510_andy-aks11510_eastus2/providers/Microsoft.Compute/disks/pvc-64ff3106-d091-45b0-ab0b-0f99c4bcb491"
I0823 14:54:03.362979       1 azure_controller_vmss.go:179] azureDisk - update(mc_andy-aks11510_andy-aks11510_eastus2): vm(aks-agentpool-60632172-vmss000006) - detach disk(pvc-64ff3106-d091-45b0-ab0b-0f99c4bcb491, /subscriptions/xxx/resourceGroups/mc_andy-aks11510_andy-aks11510_eastus2/providers/Microsoft.Compute/disks/pvc-64ff3106-d091-45b0-ab0b-0f99c4bcb491)
I0823 14:54:18.650705       1 azure_controller_vmss.go:191] azureDisk - update(mc_andy-aks11510_andy-aks11510_eastus2): vm(aks-agentpool-60632172-vmss000006) - detach disk(pvc-64ff3106-d091-45b0-ab0b-0f99c4bcb491, /subscriptions/xxx/resourceGroups/mc_andy-aks11510_andy-aks11510_eastus2/providers/Microsoft.Compute/disks/pvc-64ff3106-d091-45b0-ab0b-0f99c4bcb491) returned with <nil>
I0823 14:54:18.650763       1 azure_controller_common.go:242] azureDisk - detach disk(pvc-64ff3106-d091-45b0-ab0b-0f99c4bcb491, /subscriptions/xxx/resourceGroups/mc_andy-aks11510_andy-aks11510_eastus2/providers/Microsoft.Compute/disks/pvc-64ff3106-d091-45b0-ab0b-0f99c4bcb491) succeeded
I0823 14:54:18.650785       1 controllerserver.go:420] Trying to attach volume "/subscriptions/xxx/resourceGroups/mc_andy-aks11510_andy-aks11510_eastus2/providers/Microsoft.Compute/disks/pvc-64ff3106-d091-45b0-ab0b-0f99c4bcb491" to node "aks-agentpool-60632172-vmss000007" again
I0823 14:54:18.672831       1 azure_controller_common.go:183] Trying to attach volume "/subscriptions/xxx/resourceGroups/mc_andy-aks11510_andy-aks11510_eastus2/providers/Microsoft.Compute/disks/pvc-64ff3106-d091-45b0-ab0b-0f99c4bcb491" lun 0 to node "aks-agentpool-60632172-vmss000007".
I0823 14:54:18.672895       1 azure_controller_vmss.go:105] azureDisk - update(mc_andy-aks11510_andy-aks11510_eastus2): vm(aks-agentpool-60632172-vmss000007) - attach disk(pvc-64ff3106-d091-45b0-ab0b-0f99c4bcb491, /subscriptions/xxx/resourceGroups/mc_andy-aks11510_andy-aks11510_eastus2/providers/Microsoft.Compute/disks/pvc-64ff3106-d091-45b0-ab0b-0f99c4bcb491) with DiskEncryptionSetID()
I0823 14:54:28.968022       1 azure_controller_vmss.go:117] azureDisk - update(mc_andy-aks11510_andy-aks11510_eastus2): vm(aks-agentpool-60632172-vmss000007) - attach disk(pvc-64ff3106-d091-45b0-ab0b-0f99c4bcb491, /subscriptions/xxx/resourceGroups/mc_andy-aks11510_andy-aks11510_eastus2/providers/Microsoft.Compute/disks/pvc-64ff3106-d091-45b0-ab0b-0f99c4bcb491)  returned with <nil>
I0823 14:54:28.968067       1 controllerserver.go:428] attach volume "/subscriptions/xxx/resourceGroups/mc_andy-aks11510_andy-aks11510_eastus2/providers/Microsoft.Compute/disks/pvc-64ff3106-d091-45b0-ab0b-0f99c4bcb491" to node "aks-agentpool-60632172-vmss000007" successfully
I0823 14:54:28.968085       1 utils.go:114] GRPC response: {"publish_context":{"LUN":"0"}}
```

- Where to add such e2e test

Should add e2e test in `pre_provisioning_test.go`

After creating a disk PVC, use disksclient to attach one node, and then follow above validation steps:
https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/5a495e0427c934d275e3dd5c682d0b0beff73566/test/e2e/pre_provisioning_test.go#L79

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by landreasyan**
**What type of PR is this?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #500 

**Requirements**:
- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:
```
none
```

### CODE REVIEW

File: test/e2e/pre_provisioning_test.go
Comment (andyzhangx) [inline test/e2e/pre_provisioning_test.go:225]: remove this `"maxShares":   "5",`,  this test should be only for non-shared disk

**Review by andyzhangx — COMMENTED** [review]
the test scenario should be:
1. attach disk PV to node#1 by azure API
2. cordon node#1
3. schedule pod with same disk PV to node#2
4. make sure pod could be started successfully(with dangling error fixed)

That's a common scenario when customer has disk detach failure on node#1, and disk driver should relf-recovered.

File: test/e2e/testsuites/pre_provisioned_dangling_attach_volume_tester.go
Comment (andyzhangx) [inline test/e2e/testsuites/pre_provisioned_dangling_attach_volume_tester.go:65]: should use Azure api to attach disk, otherwise e2e test won't work since that disk PV is already attached to one node, it won't be allowed to attach to another node(blocked by k8s)

File: test/e2e/testsuites/pre_provisioned_dangling_attach_volume_tester.go
Comment (edreed) [inline test/e2e/testsuites/pre_provisioned_dangling_attach_volume_tester.go:71]: Also check that `len(nodes) > 0` here to avoid possible out-of-bounds error on line 75 if empty list is returned.

You could probably just make the check `>= 2` since we need at least two nodes for this test.

File: test/e2e/testsuites/pre_provisioned_dangling_attach_volume_tester.go
Comment (edreed) [inline test/e2e/testsuites/pre_provisioned_dangling_attach_volume_tester.go:82]: `>= 2`, but really this should make the opposite check and fail in the body of the statement instead of raising the failure in the `else` clause.

File: test/e2e/pre_provisioning_test.go
Comment (nearora-msft) [inline test/e2e/pre_provisioning_test.go:238]: Since only one pod is getting created, is this loop required here?

**Comment by landreasyan** [comment]
Need to verify the test for windows cluster that is not using vmss-s. Will squash commits before merging.

File: test/e2e/testsuites/testsuites.go
Comment (edreed) [inline test/e2e/testsuites/testsuites.go:845]: Consider `SetNodeUnschedulable` since it is setting the state of the property and not unconditionally making the node unschedulable.

File: test/e2e/testsuites/pre_provisioned_dangling_attach_volume_tester.go
Comment (andyzhangx) [inline test/e2e/testsuites/pre_provisioned_dangling_attach_volume_tester.go:76]: remove this comment since there is already logging in next line

File: test/e2e/testsuites/pre_provisioned_dangling_attach_volume_tester.go
Comment (andyzhangx) [inline test/e2e/testsuites/pre_provisioned_dangling_attach_volume_tester.go:73]: use skip test case: `ginkgo.Skip(...)`

File: test/e2e/testsuites/pre_provisioned_dangling_attach_volume_tester.go
Comment (andyzhangx) [inline test/e2e/testsuites/pre_provisioned_dangling_attach_volume_tester.go:46]: remove empty line

File: test/utils/azure/azure_helpers.go
Comment (andyzhangx) [inline test/utils/azure/azure_helpers.go:323]: remove empty line

File: test/utils/azure/azure_helpers.go
Comment (andyzhangx) [inline test/utils/azure/azure_helpers.go:304]: this would lead to error easily since it assumes there is no disk using LUN 0, what about:
1. cordon node#1
2. `azurediskDriver.ControllerPublishVolume` to attach disk to node#0 (since node#0 is cordoned)
3. uncordon node#1, cordon#0
4. create pod with that disk

By this way, you don't even need to write attach disk API

Comment (landreasyan) [reply test/utils/azure/azure_helpers.go:304]: This is the same thing I was wondering and discussed with you briefly over Teams. Using ControllerPublishVolume certainly seems like a better and easier option, but cardoning nodes will not be sufficient from what I understand. As I noted before, ControllerPublishVolume [requires NodeId](https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/master/pkg/azuredisk/controllerserver.go#L379) as a parameter and this Id is not trivial to obtain. Let me know if I am missing anything.

Comment (andyzhangx) [reply test/utils/azure/azure_helpers.go:304]: since you already use k8s api `ListNodes()`, the nodeID should be in the returned values, for vmss, its format is like `k8s-agentpool-37051515-vmss000001`, same as what you get with `kubectl get no`

Comment (landreasyan) [reply test/utils/azure/azure_helpers.go:304]: That only works for VMSets, for VMSS like I mentioned in our discussion, it returns "instance not found" error. The problem, from what I understand, is in getNodeVMSet function which misclassifies VMSS as VMSet and thus fails to find the node. If what you describe is the expected behavior, maybe I should first work on fixing the behavior of getNodeVMSet?

Comment (landreasyan) [reply test/utils/azure/azure_helpers.go:304]: Just to clarify few things and give more context, initially when the ControllerPublishVolume failed with "instance not found", I attributed it to the fact that I am passing the wrong node id and that it expect the id from nodeClient.NodeGetInfo. Now, that I succeeded running it with a different setup, VMSet instead of VMSS(my initial setup), I realized that as you mentioned, passing the node name works! So, when investigating why it doesn't work with VMSS, I saw that in my setup VMType is not specified in the config, so it defaults to VMType = "standard" instead of "vmss". When this value is set correctly to "vmss", everything works as expected. Now what I am trying to understand is, is this unique to my setup or is this a bug that needs to be addressed first? Do we always expect VMType to be provided from azure?

Comment (andyzhangx) [reply test/utils/azure/azure_helpers.go:304]: AKS does not allow VMSet and VMSS in one cluster, if it's VMSet cluster, then `VMType = "standard"` otherwise it's `VMType = "VMSS"`, will that work?

Comment (landreasyan) [reply test/utils/azure/azure_helpers.go:304]: The thing is that it is not in the same cluster. It is exclusively vmss cluster but as it doesn't find the type defined in the config(`$KUBECONFIG`), it defaults to `VMType = "standard"`, causing the problem described.

Comment (andyzhangx) [reply test/utils/azure/azure_helpers.go:304]: related to this long term issue: https://github.com/kubernetes-sigs/azuredisk-csi-driver/issues/576, while for this test case, could have a simple workaround

File: test/utils/credentials/credentials.go
Comment (edreed) [inline test/utils/credentials/credentials.go:45]: Please fix indentation.

**Review by andyzhangx — APPROVED** [review]
LGTM, could you squash all commits? thanks.

**Comment by andyzhangx** [comment]
@landreasyan run `git rebase -i HEAD~3` and follow the instructions there to squash the last two commits and you will get only one commit finally and then force push.

**Comment by landreasyan** [comment]
> @landreasyan run `git rebase -i HEAD~3` and follow the instructions there to squash the last two commits and you will get only one commit finally and then force push.

Hi Andy, all commits are already squashed except the merge commits.

**Review by andyzhangx — APPROVED** [review]
/lgtm

**Pull request merged**
